Add "Karmada + Flux" user guide (#988)
fluxcd/website: content/en/docs/use-cases/karmada.md

To start up Karmada, you can refer to [here](https://github.com/karmada-io/karmada/blob/master/docs/installation/installation.md).
If you just want to try Karmada, we recommend building a development environment by ```hack/local-up-karmada.sh```.

```sh
git clone https://github.com/karmada-io/karmada
cd karmada
hack/local-up-karmada.sh
```

After that, you will start a Kubernetes cluster by kind to run the Karmada control plane and create member clusters managed by Karmada.

```sh
kubectl get clusters --kubeconfig=/$HOME/.kube/karmada.config
```

you can use the command above to check registered clusters, and you will get similar output as follows:

```
NAME      VERSION   MODE   READY   AGE
member1   v1.23.4   Push   True    7m38s
member2   v1.23.4   Push   True    7m35s
member3   v1.23.4   Pull   True    7m27s
```

### Flux Installation 

In Karmada control plane, you only need to install Flux crds [here](https://github.com/fluxcd/flux2/blob/main/manifests/crds/kustomization.yaml) as objects delivered to member clusters.

In member clusters, you can install Flux without storing its manifests in a Git repository in member clusters for testing purposes:

```sh

> flux install

Author: Yeah, this should be ```flux install --kubeconfig=~/.kube/karmada.config --context=memberX```. I am missing. In Karmada control plane, we only apply flux crds but not need controllers to reconcile them. They are treated as resource template, not a resource instance. Based on work API [here](https://github.com/kubernetes-sigs/work-api), they will be encapsulated as a work object deliverd to member clusters and will be reconciled by controllers in member clusters finally.

> flux install

Reviewer: Ok so add the CRDs apply instructions as the first set:

```sh
 kubectl  apply -k github.com/fluxcd/flux2/manifests/crds?ref=main --kubeconfig=~/.kube/karmada.config
```

Explain what it does, then add the members commands after it.

Thanks!

```yaml
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: HelmRepository
metadata:
  name: podinfo
spec:
  interval: 1m
  url: https://stefanprodan.github.io/podinfo  
---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: podinfo
spec:
  interval: 5m
  chart:
    spec:
      chart: podinfo
      version: 5.0.3
      sourceRef:
        kind: HelmRepository
        name: podinfo
```

2. Define a PropagationPolicy that will propagate them to member clusters:

```yaml
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: helm-repo
spec:
  resourceSelectors:
    - apiVersion: source.toolkit.fluxcd.io/v1beta2
      kind: HelmRepository
      name: podinfo
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
---
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: helm-release
spec:
  resourceSelectors:
    - apiVersion: helm.toolkit.fluxcd.io/v2beta1
      kind: HelmRelease
      name: podinfo
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
```

> It means that the HelmRepository and HelmRelease created in Step 1 will be propagated to member1 and member2.

Reviewer: ```suggestion
The above configuration is for propagating the Flux objects to member1 and member2 clusters.
```

3. Apply those manifests to the Karmada-apiserver:

```sh
kubectl apply -f ../helm/ --kubeconfig=/$HOME/.kube/karmada.config
```

The output is similar to:

```

> helmrelease.helm.toolkit.fluxcd.io/podinfo created

Reviewer: How come these Flux custom resources work on this cluster? Isn't the controller plane cluster getting filled with Helm release which shouldn't be reconcile here?

---
title: Karmada + Flux
linkTitle: Karmada + Flux

> description: "How to use Karmada for distribution and management of multi-cluster helm charts."

Reviewer: ```suggestion
description: "How to use Karmada for distribution and management of multi-cluster Flux Helm releases."
```

> description: "How to use Karmada for distribution and management of multi-cluster helm releases."

Reviewer: ```suggestion
description: "How to use Karmada and Flux for distribution and management of multi-cluster Helm releases."
```

```sh
git clone https://github.com/karmada-io/karmada
cd karmada
hack/local-up-karmada.sh
```

After that, you will start a Kubernetes cluster by kind to run the Karmada control plane and create member clusters managed by Karmada.

```sh
kubectl get clusters --kubeconfig ~/.kube/karmada.config
```

you can use the command above to check registered clusters, and you will get similar output as follows:

```
NAME      VERSION   MODE   READY   AGE
member1   v1.23.4   Push   True    7m38s
member2   v1.23.4   Push   True    7m35s
member3   v1.23.4   Pull   True    7m27s
```

### Flux Installation 

In Karmada control plane, you need to install Flux crds but do not need controllers to reconcile them. They are treated as resource templates, not specific resource instances. 
Based on work API [here](https://github.com/kubernetes-sigs/work-api), they will be encapsulated as a work object deliverd to member clusters and reconciled by Flux controllers in member clusters finally.

```sh
kubectl apply -k github.com/fluxcd/flux2/manifests/crds?ref=main --kubeconfig ~/.kube/karmada.config
```

> In member clusters, you can install Flux without storing its manifests in a Git repository in member clusters for testing purposes:

Reviewer: ```suggestion
For testing purposes, we'll install Flux on member clusters without storing its manifests in a Git repository:
```

> ```sh
> flux install --kubeconfig ~/.kube/members.config --context memberX
> ```

Reviewer: ```suggestion
```sh
flux install --kubeconfig ~/.kube/members.config --context member1
flux install --kubeconfig ~/.kube/members.config --context member2
```
So people can just copy/paste.
```

Please refer to the documentations [here](../installation.md) for more ways to set up Flux in details.

{{% alert color="info" title="Tip" %}}
If you want to manage Helm releases across your fleet of clusters, Flux must be installed on each cluster.
{{% /alert %}}

## Helm release propagation

If you want to propagate Helm releases for your apps to member clusters, you can refer to the guide below.

> 1. Define a Flux `HelmRepository` manifest and a `HelmRelease` manifest in Karmada Control plane. They will serve as a resource template. 

Reviewer: ```suggestion
1. Define a Flux `HelmRepository` and a `HelmRelease` manifest in Karmada Control plane. They will serve as  resource templates. 
```

---
title: Karmada + Flux
linkTitle: Karmada + Flux
description: "How to use Karmada and Flux for distribution and management of multi-cluster Flux Helm releases."
weight: 20
---

> {{% alert color="warning" title="Disclaimer" %}}

Author: How about adding a disclaimer here?

Reviewer: Looks great, thanks!
